### PR TITLE
simplify graph signal handler

### LINKED ISSUE #3441: graphd might block at exit
**Please check the FAQ documentation before raising an issue**

**Describe the bug (__required__)**

Take a look at this issue #3434 , after reviewing GraphDaemon's code, we suspect that it may suffer from the same problem since it call `gServer->stop()` in signal handler(`gServer` is a ThriftServer instance, if the handler is triggered and scheduled to execute in one of the ThriftServer's worker thread, it may block forever):

https://github.com/vesoft-inc/nebula/blob/cd2f4a66bb86b8478f0943a02f58dd16a99f7162/src/daemons/GraphDaemon.cpp#L199-L209

**Your Environments (__required__)**

* OS: `uname -a`
* Compiler: `g++ --version` or `clang++ --version`
* CPU: `lscpu`
* Commit id (e.g. `a3ffc7d8`)

**How To Reproduce(__required__)**

1. Step 1
2. Step 2
3. Step 3

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by heroicNeZha**
#### What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### What does this PR do?
closes #3441 
simplify graphd signal handler：Set stop condition instead of waiting all workers stop

#### Which issue(s)/PR(s) this PR relates to?
In #3437 , it simplify storage SIGINT signal handler
  
#### Checklist:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the corresponding label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes:

Please confirm whether to be reflected in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/daemons/GraphDaemon.cpp [hunk 1]
```diff
@@ -166,33 +174,56 @@ int main(int argc, char *argv[]) {
...
+    gServer->setNumAcceptThreads(FLAGS_num_accept_threads);
+    gServer->setListenBacklog(FLAGS_listen_backlog);
+    if (FLAGS_enable_ssl || FLAGS_enable_graph_ssl) {
+      gServer->setSSLConfig(nebula::sslContextConfig());
+    }
+    setupThreadManager();
+    serverStatus_.store(STATUS_RUNNING);
+    FLOG_INFO("Starting nebula-graphd on %s:%d\n", localhost.host.c_str(), localhost.port);
+    try {
+      gServer->serve();  // Blocking wait until shut down via gServer->stop()
```
Comment (Shylock-Hg) [inline src/daemons/GraphDaemon.cpp:198]: I think you just need call `serve` in this thread.

Comment (heroicNeZha) [reply src/daemons/GraphDaemon.cpp:198]: When I call `serve()`, the deamon will be blocked here, I don’t know how to ensure that the `stop` call is not a worker thread

File: src/daemons/GraphDaemon.cpp [hunk 2]
```diff
@@ -43,6 +46,11 @@ extern Status setupBreakpad();
 DECLARE_string(flagfile);
 DECLARE_bool(containerized);
 
+enum ServiceStatus { STATUS_UNINITIALIZED = 0, STATUS_RUNNING = 1, STATUS_STOPPED = 2 };
+std::atomic<ServiceStatus> serverStatus_{STATUS_UNINITIALIZED};
+std::mutex muStop_;
+std::condition_variable cvStop_;
```
Comment (yixinglu) [inline src/daemons/GraphDaemon.cpp:52]: Prefer to provide a class to handle this signal process

Comment (heroicNeZha) [reply src/daemons/GraphDaemon.cpp]: done

File: src/graph/service/GraphServer.cpp [hunk 3]
```diff
@@ -24,7 +24,7 @@ bool GraphServer::start() {
   int numThreads =
       FLAGS_num_worker_threads > 0 ? FLAGS_num_worker_threads : gServer_->getNumIOWorkerThreads();
   std::shared_ptr<apache::thrift::concurrency::ThreadManager> threadManager(
-      PriorityThreadManager::newPriorityThreadManager(numThreads, false /*stats*/));
```
Comment (Shylock-Hg) [inline src/graph/service/GraphServer.cpp]: Why change this?

Comment (heroicNeZha) [reply src/graph/service/GraphServer.cpp]: After updating third party 3.0，newPriorityThreadManager remove second argument

File: src/graph/service/GraphServer.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+ private:
+  HostAddr localHost_;
+
+  std::shared_ptr<folly::IOThreadPoolExecutor> ioThreadPool_;
+  std::shared_ptr<apache::thrift::concurrency::ThreadManager> workers_;
+  std::unique_ptr<apache::thrift::ThriftServer> gServer_;
+  std::unique_ptr<std::thread> graphThread_;
+
+  enum ServiceStatus { STATUS_UNINITIALIZED = 0, STATUS_RUNNING = 1, STATUS_STOPPED = 2 };
```
Comment (yixinglu) [inline src/graph/service/GraphServer.h:39]: enum ServiceStatus : uint8_t {}

File: src/graph/service/GraphServer.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  void notifyStop();
+
+  void waitUntilStop();
+
+ private:
+  HostAddr localHost_;
+
+  std::shared_ptr<folly::IOThreadPoolExecutor> ioThreadPool_;
+  std::shared_ptr<apache::thrift::concurrency::ThreadManager> workers_;
+  std::unique_ptr<apache::thrift::ThriftServer> gServer_;
```
Comment (yixinglu) [inline src/graph/service/GraphServer.h:36]: thriftServer_

File: src/graph/service/GraphServer.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+#include "GraphServer.h"
+
+#include <memory>
+
+#include "graph/service/GraphFlags.h"
+#include "graph/service/GraphService.h"
+namespace nebula {
+namespace graph {
+
+GraphServer::GraphServer(HostAddr localHost) : localHost_(localHost) {}
```
Comment (yixinglu) [inline src/graph/service/GraphServer.cpp:14]: std::move(localHost)

File: src/daemons/GraphDaemon.cpp [hunk 7]
```diff
@@ -153,47 +152,23 @@ int main(int argc, char *argv[]) {
...
-  FLOG_INFO("Starting nebula-graphd on %s:%d\n", localhost.host.c_str(), localhost.port);
-  try {
-    gServer->serve();  // Blocking wait until shut down via gServer->stop()
-  } catch (const std::exception &e) {
-    FLOG_ERROR("Exception thrown while starting the RPC server: %s", e.what());
+  gServer = std::make_unique<nebula::graph::GraphServer>(localhost);
+
+  if (!gServer->start()) {
+    LOG(ERROR) << "The graph server start failed";
+    gServer->stop();
```
Comment (yixinglu) [inline src/daemons/GraphDaemon.cpp:166]: Need not to call stop explicitly

File: src/daemons/GraphDaemon.cpp [hunk 8]
```diff
@@ -29,20 +29,19 @@ using nebula::fs::FileUtils;
...
 static void printHelp(const char *prog);
-static void setupThreadManager();
 #if defined(__x86_64__)
 extern Status setupBreakpad();
 #endif
 
 DECLARE_string(flagfile);
 DECLARE_bool(containerized);
 
+std::unique_ptr<nebula::graph::GraphServer> gServer;
```
Comment (yixinglu) [inline src/daemons/GraphDaemon.cpp:43]: don't forget to update the Standalone daemon introduced in #3310

File: src/graph/service/GraphServer.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    }
+  }
+
+  gServer_->stop();
+
+  graphThread_->join();
+}
+
+void GraphServer::notifyStop() {
+  std::unique_lock<std::mutex> lkStop(muStop_);
```
Comment (yixinglu) [inline src/graph/service/GraphServer.cpp:84]: Replace this lock with atomic CAS operation.

**Review by yixinglu — COMMENTED** [review]
A little complicated implementation, could use `init` and `start` function to do this. And need not to spawn a new thread for thrift server. you can talk with me offline.

File: src/graph/service/GraphServer.cpp (hunk 3 above)
Comment (CPWstatic) [reply src/graph/service/GraphServer.cpp]: This argument affects the performance a lot, we must know how it behaves now. And make sure if it has another way to set the value. A benchmark must done for this change.

Comment (heroicNeZha) [reply src/graph/service/GraphServer.cpp]: https://github.com/facebook/fbthrift/commit/bfb527c7b254230cf5ea6ceb818c8ce57a3f9a84?diff=split
Through this commit，thrift removed stats, and the function is the same as setting it to false

File: src/graph/service/GraphServer.cpp (hunk 9 above)
Comment (heroicNeZha) [reply src/graph/service/GraphServer.cpp:83]: if remove this lock , it will be thread unsafe

**Review by CPWstatic — DISMISSED** [review]

File: src/graph/service/GraphServer.cpp [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+void GraphServer::stop() {
+  if (serverStatus_.load() == ServiceStatus::STATUS_STOPPED) {
+    LOG(INFO) << "The graph server has been stopped";
+    return;
+  }
+
+  ServiceStatus serverExpected = ServiceStatus::STATUS_RUNNING;
+  serverStatus_.compare_exchange_strong(serverExpected, STATUS_STOPPED);
```
Comment (Shylock-Hg) [inline src/graph/service/GraphServer.cpp:97]: Don't wait it to running?

**Review by yixinglu — APPROVED** [review]

File: src/graph/service/GraphServer.cpp (hunk 10 above)
Comment (heroicNeZha) [reply src/graph/service/GraphServer.cpp:97]: it wait in thriftServer.stop

**Review by Shylock-Hg — APPROVED** [review]

**Pull request merged**